src/decoder.c

```diff
@@ -0,0 +1,393 @@
+            case ED448_BIT_SIZE:
+                data_type = ED448;
+                break;
+            case ED25519_BIT_SIZE:
+                data_type = ED25519;
+                break;
+            default:
+                return RET_OSSL_ERR;
+            }
+            break;
+        default:
+            return RET_OSSL_ERR;
+        }
+        p11prov_obj_to_store_reference(obj, &reference, &reference_len);
+        if (!reference) {
+            return RET_OSSL_ERR;
+        }
+        break;
+    case CKO_CERTIFICATE:
+        object_type = OSSL_OBJECT_CERT;
+        data_type = "CERTIFICATE";
+        cert = p11prov_obj_get_attr(obj, CKA_VALUE);
+        if (!cert) {
+            return RET_OSSL_ERR;
+        }
+        break;
+    default:
+        return RET_OSSL_ERR;
+    }
+
+    params[0] = OSSL_PARAM_construct_int(OSSL_OBJECT_PARAM_TYPE, &object_type);
+    params[1] = OSSL_PARAM_construct_utf8_string(OSSL_OBJECT_PARAM_DATA_TYPE,
+                                                 (char *)data_type, 0);
+    if (reference) {
+        /* giving away the object by reference */
+        params[2] = OSSL_PARAM_construct_octet_string(
+            OSSL_OBJECT_PARAM_REFERENCE, reference, reference_len);
+    } else if (cert) {
+        params[2] = OSSL_PARAM_construct_octet_string(
+            OSSL_OBJECT_PARAM_DATA, cert->pValue, cert->ulValueLen);
+    } else {
+        return RET_OSSL_ERR;
+    }
+    params[3] = OSSL_PARAM_construct_end();
+    return RET_OSSL_OK;
+}
+
+static bool decoder_ctx_accepts_decoded_object(P11PROV_DECODER_CTX *ctx)
+{
+    return (!ctx->invalid) && (!ctx->object);
+}
+
+static void decoder_ctx_object_free(struct p11prov_decoder_ctx *ctx)
+{
+    if (ctx && ctx->object) {
+        p11prov_obj_free(ctx->object);
+        ctx->object = NULL;
+    }
+}
+
+static void *p11prov_decoder_newctx(void *provctx)
+{
+    P11PROV_CTX *cprov;
+    P11PROV_DECODER_CTX *dctx;
+    cprov = provctx;
+    dctx = OPENSSL_zalloc(sizeof(P11PROV_DECODER_CTX));
+    if (!dctx) {
+        return NULL;
+    }
+
+    dctx->provctx = cprov;
+    return dctx;
+}
+
+static void p11prov_decoder_freectx(void *inctx)
+{
+    P11PROV_DECODER_CTX *ctx = inctx;
+
+    decoder_ctx_object_free(ctx);
+    OPENSSL_clear_free(ctx, sizeof(P11PROV_DECODER_CTX));
+}
+
+static CK_RV p11prov_decoder_ctx_store_obj(void *pctx, P11PROV_OBJ *obj)
+{
+    P11PROV_DECODER_CTX *ctx = pctx;
+
+    if (!decoder_ctx_accepts_decoded_object(ctx)) {
+        P11PROV_raise(ctx->provctx, CKR_GENERAL_ERROR,
+                      "decoder context does not accept any objects");
+        ctx->invalid = 1;
+        decoder_ctx_object_free(ctx);
+        p11prov_obj_free(obj);
+        return CKR_GENERAL_ERROR;
+    }
+
+    P11PROV_debug("Adding object (handle:%lu)", p11prov_obj_get_handle(obj));
+    ctx->object = obj;
+
+    return CKR_OK;
+}
+
+static CK_RV p11prov_decoder_load_obj(P11PROV_DECODER_CTX *ctx,
```

Comment:
Why have you re-implemented store logic in here instead of just calling p11prov_store_open() followed by p11prov_store_load() ?

I do not think replicating the store logic outside of the store is good for maintainability.

[Of course you can have a helper wrapper function in src/store.c so that we do not need to export all of these static functions individually]

Reply:
Re-implementation is gone. p11prov_store_direct_fetch() is exposed by the store.c to do a open+load.